### PR TITLE
Add proper shape checking to torch.cat

### LINKED ISSUE #4071: torch.cat doesn't do proper shape checking, which can lead to incorrect gradients
This is a *very* sneaky behaviour so I will call it a bug as it can easily break your code.
I paste here a self-contained example. Even though `axes1` is equivalent to `axes2`, `axes2` gives wrong gradients.
```python
from torch.autograd import Variable
from torch import autograd
import torch
import numpy as np

def norm_vec(v):
    return torch.sqrt(torch.pow(v, 2).sum())

coo = Variable(torch.from_numpy(np.array([[3, 4, 2],])).float(), requires_grad=True)
axes1 = torch.stack([coo[0] / norm_vec(coo[0]), Variable(torch.zeros(3), requires_grad=True), Variable(torch.zeros(3), requires_grad=True)], dim=1)
axes2 = torch.stack([coo[0] / norm_vec(coo[0]), Variable(torch.zeros(3, 1), requires_grad=True), Variable(torch.zeros(3, 1), requires_grad=True)], dim=1).squeeze()
new1 = torch.matmul(coo, axes1)
new2 = torch.matmul(coo, axes2)
print(autograd.grad(new1[0, 0], coo, create_graph=True), autograd.grad(new2[0, 0], coo, create_graph=True))
```

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #4071.

`THTensor_(catArray)` and `THCTensor_(catArray)` do some strange thing where the shapes of the tensors don't have to be the same in some dimensions in some cases. This adds full shape checking.

Non-empty tensor arguments to torch.cat must all have the same shape, except in the cat dimension.

### Test Plan
New unit tests:
- CPU shape check test
- CUDA shape check test
- CUDA torch.cat general test to match CPU's test

### CODE REVIEW

File: aten/src/TH/generic/THTensorMath.c [hunk 1]
```diff
@@ -2809,116 +2809,114 @@ void THTensor_(cat)(THTensor *r_, THTensor *ta, THTensor *tb, int dimension)
   THTensor_(catArray)(r_, inputs, 2, dimension);
 }
 
+void THTensor_(check_shape_except_dim)(THTensor *first, THTensor *second, int dimension);
+inline void THTensor_(check_shape_except_dim)(THTensor *first, THTensor *second, int dimension)
+{
+  int64_t first_dims = first->nDimension;
+  int64_t second_dims = second->nDimension;
+  THArgCheck(first_dims == second_dims, 0,
+      "Tensors must have same number of dimensions: got %d and %d",
```
Comment (colesbury) [inline aten/src/TH/generic/THTensorMath.c]: It looks like the format specifiers don't match types: `%d` vs. `int64_t`

**Comment by ezyang** [comment]
I hate to ask this now that the code is all written, but is there a compelling reason why this couldn't have been implemented in ATen? (I guess, no easy way to insert the checks into the generated code?)

**Comment by zou3519** [comment]
Good point, I hadn't thought about implementing the shape checks in aten. I think it shouldn't be too bad to do that: I could rename `cat` in the cwrap to something and implement a native function that wraps the renamed `cat` along with performing size checks.

The two downsides I see with this approach are that:
- tensors don't benefit from the size check (only Variables, since size check is in aten). When tensors/Variables get merged this won't be a problem
- we'd be doing an extra pass through the list of tensors to cat. Right now the size of the final tensor is computed at the same time as the size checks; separating the two would mean two loops, one for the size checks (in aten), and another to compute the size of the final tensor (in TH). This probably doesn't matter.

Alternatively I could just rewrite `cat` completely as a native function

**Comment by soumith** [comment]
we should keep them in TH/THC. there's no upside (now that the code is written) to do this in ATen.
